compaction/compaction.cc

```diff
@@ -433,7 +433,7 @@ class compaction {
     std::unordered_set<shared_sstable> _new_partial_sstables;
     std::vector<shared_sstable> _new_unused_sstables;
     std::vector<shared_sstable> _all_new_sstables;
-    lw_shared_ptr<sstable_set> _compacting;
+    optimized_optional<sstable_set> _compacting;
     const sstables::compaction_type _type;
```

Comment:
I think the patch title "sstable_set: allow updating in-place" is misleading. We're laying the ground for that, but this patch is changing sstable_set to be a shared pointer instead of value object.

I think we don't want to update sstable_set in place, because it should remain a snapshot.

Reply:
Doing so creates complexity in taking a snapshot of the sstable_set in the for of a `lw_shared_ptr<sstable_set>` that we just got used to, but I don't think it's any less complicated than pushing the shared ptr one layer down, into the sstable_set itself, and having the sstable_set_impl contain the value object.

Plus, it forces us to recalculate the table's compound sstable_set when updating even a single sstable_set, the way compound_sstable_set is built today.

To prevent the latter, I think that we should move towards a state-less compound_sstable_set that will walk over the table's sstable_set:s dynamically.  FWIW, its incremental selector can take a snapshot of the incremental selectors of all sets when built, very much like it does today (and maybe even that is not required as you pointed out in https://github.com/scylladb/scylladb/pull/14043#discussion_r1315195879).

Reply:
> pushing the shared ptr one layer down

This is fine

> it forces us to recalculate the table's compound sstable_set when updating even a single sstable_set, the way compound_sstable_set is built today.

Maybe we shouldn't keep per-table sstable set all the time and build it on demand instead?

Changing sstable_set underneath our feet violates immutability which causes correctness issues.

Reply:
A copy of an sstable_set is immutable, since only the original object will change in place. Very much like a copy of a `lw_shared_ptr<sstable_set>` that we keep today.

Reply:
> A copy of an sstable_set is immutable, since only the original object will change in place. Very much like a copy of a lw_shared_ptr<sstable_set> that we keep today.

Just after this change, yes. But not when you make compound_sstable_set state-less. After that, copying table::_sstables is no longer taking a snapshot. That's why you had to clone the sstable_set deeply in table::sstables_as_snapshot_source(), to restore snapshot semantics. In every place which currently assumes snapshot semantics from lw_shared_ptr<sstable_set>, we would have to do that. How do we know it's the only place?